transports/webrtc/src/upgrade.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::channel::oneshot;
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use log::{debug, error, trace};
+use webrtc::api::APIBuilder;
+use webrtc::data_channel::data_channel_init::RTCDataChannelInit;
+use webrtc::dtls_transport::dtls_role::DTLSRole;
+use webrtc::peer_connection::configuration::RTCConfiguration;
+use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;
+use webrtc_data::data_channel::DataChannel as DetachedDataChannel;
+use webrtc_ice::udp_mux::UDPMux;
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::connection::Connection;
+use crate::error::Error;
+use crate::sdp;
+use crate::transport;
+
+pub struct WebRTCUpgrade {}
+
+impl WebRTCUpgrade {
+    pub async fn new(
+        udp_mux: Arc<dyn UDPMux + Send + Sync>,
+        config: RTCConfiguration,
+        addr: Multiaddr,
+    ) -> Result<Connection<'static>, Error> {
```

Comment:
If we are not storing any state and don't return `Self`, we can probably do away with this struct!

Reply:
Is this code even used anywhere? This seems strangely similar to the `WebRTCTransport::do_dial` function.

Reply:
> Is this code even used anywhere? This seems strangely similar to the `WebRTCTransport::do_dial` function.

Yes, see `WebRTCListenStream::poll_next`. It's similar to `do_dial`, but does some things a bit differently.